### PR TITLE
Refactor write buffer in lookupsIO

### PR DESCRIPTION
**Pull request by dcoutts**
See the individual patches for more detailed descriptions.

### CODE REVIEW

File: src/Database/LSMTree/Internal/WriteBufferBlobs.hs [hunk 1]
```diff
@@ -123,6 +124,9 @@ data BlobFileState m h =
      , blobFilePointer :: !(FilePointer m)
      }
 
+instance NFData (WriteBufferBlobs m h) where
+  rnf (WriteBufferBlobs _ _ _) = ()
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal/WriteBufferBlobs.hs:129]:
```suggestion
instance NFData (WriteBufferBlobs m h) where
  rnf (WriteBufferBlobs _ b c) = rnf b `seq` rnf c
```

File: src/Database/LSMTree/Normal.hs [hunk 2]
```diff
@@ -289,17 +290,16 @@ lookups ::
...
-    V.mapStrict (bimap Internal.deserialiseValue BlobRef) <$!>
-    Internal.lookups const (V.map Internal.serialiseKey ks) th toNormalLookupResult
-
-toNormalLookupResult :: Maybe (Entry.Entry v b) -> LookupResult v b
-toNormalLookupResult = \case
-    Just e -> case e of
-      Entry.Insert v            -> Found v
-      Entry.InsertWithBlob v br -> FoundWithBlob v br
-      Entry.Mupdate _           -> error "toNormalLookupResult: Mupdate unexpected"
+    V.map toLookupResult <$>
```
Comment (jorisdral) [inline src/Database/LSMTree/Normal.hs:293]: Is no longer using `mapStrict` intentional?

File: test/Test/Database/LSMTree/Internal/Lookup.hs [hunk 3]
```diff
@@ -288,38 +290,65 @@ prop_inMemRunLookupAndConstruction dat =
...
+         -> [InMemLookupData SerialisedKey SerialisedValue SerialisedBlob]
+         -> (   WB.WriteBuffer
+             -> WBB.WriteBufferBlobs IO h
+             -> V.Vector (Run.Run IO (Handle h))
+             -> IO a)
+         -> IO a
+withRuns hfs _ [] action = do
+    wbblobs <- WBB.new hfs (FS.mkFsPath ["wbblobs"])
+    r <- action WB.empty wbblobs V.empty
+    WBB.removeReference wbblobs
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Internal/Lookup.hs:338]: We should probably use `bracket` here

File: test/Test/Database/LSMTree/Internal/Lookup.hs [hunk 4]
```diff
@@ -288,38 +290,65 @@ prop_inMemRunLookupAndConstruction dat =
...
+        [ mkRunFromSerialisedKOps hfs hbio fsPaths runData
+        | (i, InMemLookupData{runData}) <- zip [1..] rundats
+        , let fsPaths = RunFsPaths (FS.mkFsPath []) (RunNumber i)
         ]
-    lookupss = V.fromList $ concatMap lookups dats
-    resolveV = \(SerialisedValue v1) (SerialisedValue v2) -> SerialisedValue (v1 <> v2)
+    r <- action wb wbblobs runs
+    V.mapM_ Run.removeReference runs
+    WBB.removeReference wbblobs
+    return r
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Internal/Lookup.hs]: Same

File: test/Test/Database/LSMTree/Internal/Lookup.hs [hunk 5]
```diff
@@ -288,38 +290,65 @@ prop_inMemRunLookupAndConstruction dat =
...
+        arenaManager
+        resolveV
+        wb wbblobs
+        runs
+        (V.map Run.runFilter runs)
+        (V.map Run.runIndex runs)
+        (V.map Run.runKOpsFile runs)
+        keys
+    -- TODO: compare blobs: we can do this now we implemented blob retrieval.
+    pure $ opaqueifyBlobs modelres === opaqueifyBlobs realres
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Internal/Lookup.hs]: You could remove the `opaqueifyBlobs` in this PR

File: test/Test/Database/LSMTree/Internal.hs [hunk 6]
```diff
@@ -173,17 +173,16 @@ prop_interimOpenTable dat = ioProperty $
...
           close th'
           -- TODO: checking lookups is a simple check, but we could have stronger
           -- guarantee. For example, we might check that the internal structures
           -- match.
           --
-          -- TODO: remove opaqueifyBlobs once blobs are implemented.
           pure $ tabulate "Number of runs snapshotted" [show numRunsSnapped]
-              $ (Test.opaqueifyBlobs lhs === Test.opaqueifyBlobs rhs)
+               $ lhs === rhs
   where
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Internal.hs:186]: Address the TODO below that is on `runDataWithoutBlobs`

Comment (jorisdral) [reply test/Test/Database/LSMTree/Internal.hs:186]: There is also one in `prop_roundtripCursor`

Comment (jorisdral) [reply test/Test/Database/LSMTree/Internal.hs:186]: There is also a TODO regarding blob references in `Test.Database.LSMTree.Internal.RunReaders.arbitraryWithVars`

Comment (jorisdral) [reply test/Test/Database/LSMTree/Internal.hs:186]: And `Test.Database.LSMTree.Normal.Examples.tests`

Comment (jorisdral) [reply test/Test/Database/LSMTree/Internal.hs:186]: I grepped for TODOs mentioning blobs and I listed the ones we still have to remove above. It's not really all in scope for this PR, but they are probably small enough changes that you can make them in this PR.

**Review by jorisdral — CHANGES_REQUESTED** [review]
Just some small suggestions, the rest looks good

File: src/Database/LSMTree/Normal.hs (hunk 2 above)
Comment (dcoutts) [reply src/Database/LSMTree/Normal.hs:293]: Yes. The `V.map` could fuse in a caller.

File: test/Test/Database/LSMTree/Internal.hs (hunk 6 above)
Comment (dcoutts) [reply test/Test/Database/LSMTree/Internal.hs:186]: Thanks for the audit. I'll have a look at the others.

Comment (jorisdral) [reply test/Test/Database/LSMTree/Internal.hs:184]: Looks good!

**Review by jorisdral — APPROVED** [review]

**Comment by dcoutts** [comment]
I've removed the patch for enabling blobs in various tests since it doesn't need to be in this PR at all, and some of them need fixing. I'll open a separate PR for those.

**Pull request merged**